spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/AbstractHazelcastMessageProducer.java

```diff
@@ -0,0 +1,200 @@
+
+import java.io.Serializable;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EventObject;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.hazelcast.core.AbstractIMapEvent;
+import com.hazelcast.core.DistributedObject;
+import com.hazelcast.core.EntryEvent;
+import com.hazelcast.core.EntryListener;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.MapEvent;
+import com.hazelcast.core.MultiMap;
+
+import reactor.util.StringUtils;
+
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.hazelcast.common.CacheListeningPolicyType;
+import org.springframework.integration.hazelcast.common.HazelcastIntegrationDefinitionValidator;
+import org.springframework.integration.hazelcast.common.HazelcastIntegrationUtils;
+import org.springframework.integration.hazelcast.config.HazelcastLocalInstanceHandler;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.Assert;
+
+/**
+ * Hazelcast Base Event-Driven Message Producer.
+ *
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ *
+ */
+public abstract class AbstractHazelcastMessageProducer extends MessageProducerSupport implements DisposableBean {
+
+	private final DistributedObject distributedObject;
+
+	private String cacheEventTypes;
+
+	private CacheListeningPolicyType cacheListeningPolicy;
+
+	private String hazelcastRegisteredEventListenerId;
+
+	public AbstractHazelcastMessageProducer(DistributedObject distributedObject) {
+		Assert.notNull(distributedObject, "cache must not be null");
+		this.distributedObject = distributedObject;
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+		HazelcastIntegrationDefinitionValidator.validateCacheEventsByDistributedObject(
+				getDistributedObject(), this.getCacheEventTypeSet());
+	}
+
+	protected Set<String> getCacheEventTypeSet() {
+		return StringUtils.commaDelimitedListToSet(getCacheEventTypes());
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		HazelcastIntegrationUtils.shutdownAllHazelcastInstances();
+	}
+
+	protected DistributedObject getDistributedObject() {
+		return distributedObject;
+	}
+
+	protected String getCacheEventTypes() {
+		return cacheEventTypes;
+	}
+
+	public void setCacheEventTypes(String cacheEventTypes) {
+		this.cacheEventTypes = cacheEventTypes;
+	}
+
+	protected CacheListeningPolicyType getCacheListeningPolicy() {
+		return cacheListeningPolicy;
+	}
+
+	public void setCacheListeningPolicy(CacheListeningPolicyType cacheListeningPolicy) {
+		this.cacheListeningPolicy = cacheListeningPolicy;
+	}
+
+	protected String getHazelcastRegisteredEventListenerId() {
+		return hazelcastRegisteredEventListenerId;
+	}
+
+	public void setHazelcastRegisteredEventListenerId(String hazelcastRegisteredEventListenerId) {
```

Comment:
`protected`. It would be bad to allow end-user to mutate on this local property.